### ISSUE TITLE
Add new Windows test to run system command via PowerShell called "powershellcommand"

### ISSUE THREAD

**Issue by vanderpol** [OVAL-Community/OVAL#151]
Abstract
Add a Windows test called "powershellcommand" which supports running any arbitrary command or contents of a powershell script on the target system. This feature should only be used if the content is trusted, ideally digitally signed either at the benchmark level, OVAL document level, or at the object level.

This allows for many more system configuration and vulnerability tests to be performed with OVAL without making the language overly large by adding numerous windows commands to OVAL. This also will allow for security personal not well versed with OVAL to be able to create OVAL content. It should also help to simplify some of the overly complex binary trees of tests that has plagued OVAL in the past.

Link to Proposal
Once a proposal has been put into pull request form, add a link to the PR here. If and as alternate proposals or objections are added they should be linked here as well.

Additional context
Add any other context or screenshots about the enhancement.

**Comment by wilsonbryanl** [comment]
[windows_powershellcommand_v1_20241024.zip](https://github.com/user-attachments/files/17509590/windows_powershellcommand_v1_20241024.zip)
Version 1.0 Of PowerShell_Command proposal

**Comment by wilsonbryanl** [comment]
Proposed OVAL Addition Powershell_Command Test 
 
October 24, 2024
Version: 1.0
 
1. Background 
 
OVAL test types have been relatively specific in their targeted functions and, typically, have been applicable to a specific system, sub-system, or application on a target machine.  The time and effort required to add new tests to the OVAL specific and the relative scarcity of resources impede efficient development and support the specification.  To counter that dynamic and to broadly expand the breadth of OVAL tests, in general, we propose two new OVAL tests, one each for the UNIX and Windows families, the Unix shell_comand test and the Windows powershell_command test.  Each is much broader in their application than the previously narrow tests currently available in OVAL allowing a content author to use the full capability of the Unix shell (e.g. bash) or the Windows powershell. 

This document proposes the Windows powershell_command test, specifically.  A second proposal covers the Unix shell_command test.

Since this test requires the running of code supplied by content and since SCAP applications commonly run with elevated privileges, this places significant responsibility on the content author to do no harm.  This also requires that any content stream that employs one of these tests MUST be from a known and trusted source and be digitally signed.

The advantages of these proposed tests include: 
•	dramatic flexibility in OVAL test design strategy.
•	simple content and results structure.
•	substantially reduced content development and testing resources.
•	consistent the existing OVAL processing model.
•	while initially proposed for the Unix and Windows Operating System families, this approach could later be easily extended to support other targets.

Once these tests are added, there will be a significant reduction in the need for other new OVAL tests and the resources required to develop and implement them.

**Comment by vanderpol** [comment]
@OVAL-Community/oval-board-members we are looking for your high level feedback on this proposal before we start the PR process, thanks in advance!

**Comment by solind** [comment]
Hi @vanderpol, just curious, have you considered the possibility of just using the existing `cmdlet_test` for this purpose?  It's already pretty flexible, particularly if you consider that it could be used to run custom cmdlets, provided they're installed and available beforehand.

**Comment by vanderpol** [comment]
@solind yes we have considered this, and in looking through DISA's STIGs for rules to automate, many of the ones that we found could be automated with cmdlet, but there are several that require running windows commands, which is not possible using cmdlet.  Our goal for adding a powershell test gives the ability to run any Windows system command (basically anything from a cmd prompt, and anything from a powershell prompt).  Given that we seem to have 'one shot' to expand OVAL functionality and make it future proof, we wanted to go for as much functionality as possible.  We plan to continue using the existing cmdlet test for most powershell related queries, but using this new test for things beyond "Get-" type data.

**Comment by solind** [comment]
OK, but, if you had a platform-independent `shellcommand_test`, couldn't you use that to run non-cmdlet Windows commands?

**Comment by vanderpol** [comment]
@solind, we could but would it be clear to content authors what shell that was running in on Windows?  command shell vs powershell matter, and I guess we could make the independent test include platform specific documentation on how it's intended to work on Windows vs Linux, but that makes it pretty obvious that's it's not really independent in more than name only.  We don't have any test content that uses powershell beyond cmdlet right now, but we want to provide that capability for some automation going forward.  

Once again we definitely good do this all via an independent test, but splitting them apart is very little effort, and provides content author clarity on what to expect, and allows the tests to drift apart on Windows to *nix later on if needed.

**Comment by solind** [comment]
@vanderpol I just always assumed it must be `cmd.exe`, because I'm old enough to remember a time that there was no such thing as Powershell!

You're right, some commands behave a little differently when running inside PowerShell -- at least, the way you may have to quote the arguments can change.

Clearly Unix and Windows commands are not interchangeable, but it's also the case that not all commands work the same across Unix platforms, or even across different flavors of Linux.  Nor is there any way to indicate which Unix shell the command should be run in (e.g., Bourne, bash, zsh) -- that can be just as material!  I believe Joval always used `/bin/sh`, but maybe it was `/usr/bin/env`; I'm not positive.

So, perhaps you should add a "shell" entity for the command.

**Comment by solind** [comment]
I'd add, these fundamentals (command shell, stdout, stderr, [stdin], exit code) have been essentially unchanged at least since the early 80s.  Young people don't even seem to know they exist at all.  I don't think they're likely to change in the future.

**Comment by vanderpol** [comment]
@solind, I was just going to suggest a shell entity type for the command, and then I could see where this could be merged into an independent test.  

Potential entity types
cmd
powershell
sh
bash
ksh

If the shell selected in the content does note exist, the test would need to error.   On windows cmd is 'everywhere' but powershell may not be, at least with older versions, I see now that you can't remove it from Windows 11 etc..  

For Linux /bin/sh is definitely the safest bet, but bash is nearly as common now, @johnulmer-oval may want to join in this conversation, as I'm not that savvy on Linux, and I always use bash.

**Comment by vanderpol** [comment]
Definitely an enumeration IMHO, where we could document what to expect, and what platforms each supports.  See the unix shell test for some cross proposal discussions there about a separate element vs an attribute on the command element.

**Comment by A-Biggs** [comment]
I think the list should align with the list of languages for which SCE scripts support at a minimum.  From the XCCDF specification:
A script written in the given language. Executing the script should bring the target system into compliance with the rule. The following languages are pre-defined:

> - sh – Bourne shell
> - csh – C Shell
> - perl – Perl
> - batch – Windows batch script
> - python – Python and all Python-based scripting languages
> - vbscript – Visual Basic Script (VBS)
> - javascript – Javascript (ECMAScript, Jscript)
> - tcl – Tcl and all Tcl-based scripting languages
We could add PowerShell to that list, and maybe a few others like Ruby, Jython, and Apple Script.

I believe one could use `/bin/sh -c` to run a script on Unix based systems and that would use the shebang at the top of a file to determine the interpreter to use to run the script, so that might be one way to "extend" the supported shell command types on unix.  Not sure if we want to use that to remove types from the list, as some of those types can run on windows and unix, and I don't think the shebang behaviour works on Windows, so we would need the types like perl and python there to be able to run those types on Windows.

**Comment by vanderpol** [comment]
@A-Biggs I think this goes way beyond the scope we were pondering for this effort.  Our goal is to provide a simple to use method to run system commands gather data not currently gatherable in OVAL without many new tests.  Adding entire programming languages beyond shell/powershell is opening up an even larger can of worms for debate.

A shell, and a programming language to me seem like different scopes, but maybe I'm over simplifying the situation.

**Comment by johnulmer-oval** [comment]
> I think the list should align with the list of languages for which SCE scripts support at a minimum. From the XCCDF specification: A script written in the given language. Executing the script should bring the target system into compliance with the rule. The following languages are pre-defined:
> 
> > * sh – Bourne shell
> > * csh – C Shell
> > * perl – Perl
> > * batch – Windows batch script
> > * python – Python and all Python-based scripting languages
> > * vbscript – Visual Basic Script (VBS)
> > * javascript – Javascript (ECMAScript, Jscript)
> > * tcl – Tcl and all Tcl-based scripting languages
> >   We could add PowerShell to that list, and maybe a few others like Ruby, Jython, and Apple Script.
> 
> I believe one could use `/bin/sh -c` to run a script on Unix based systems and that would use the shebang at the top of a file to determine the interpreter to use to run the script, so that might be one way to "extend" the supported shell command types on unix. Not sure if we want to use that to remove types from the list, as some of those types can run on windows and unix, and I don't think the shebang behaviour works on Windows, so we would need the types like perl and python there to be able to run those types on Windows.

That is interesting and, in general, I like the idea....  however, it seems like we'd be trying to implement a universal code interpreter in SCAP processors.  The number of underlying layers of requirements that could be brought in would be challenging. 

Unfortunately, we are on a very compressed schedule.  That is part of why we have made the proposals relatively narrow using the unix shell (defaulting to bash) and the windows powershell.  If we could make those two available to content authors, much of what is not automatable in lots of content would be easily and quickly automated.  Lots of bang for the buck.

I'd like to see the two proposals (maybe collapsed into one) added in the next few weeks.  If there is interest in a wider treatment of the idea of running code via OVAL to gather sys chars via SCE or similar, we could pursue that later.

**Comment by solind** [comment]
We're not really talking about running a whole script here, like SCE does.  This test is just for when you want to run a relatively simple one-line command.

With SCE, there are references you can make to external documents or streams containing the scripts themselves.  There's nowhere to put a script in an OVAL definitions file.

**Comment by vanderpol** [comment]
Closing this issue, no change required as it is merging with #150
